src/sorting/countingsort.lua

```diff
@@ -0,0 +1,58 @@
+function counting_sort(arr, key_function)
+     -- If no key_function is provided, use identity function
+     key_function = key_function or function(x) return x end
+
+     -- Step 1: Find the range of keys (min_key and max_key)
+     local min_key, max_key = math.huge, -math.huge
+     for i = 1, #arr do
+         local key = key_function(arr[i])
+         if key < min_key then
+             min_key = key
+         end
+         if key > max_key then
+             max_key = key
+         end
+     end
+
+     -- Step 2: Initialize the count array
+     local count = {}
+     for i = min_key, max_key do
```

Comment:
You get a problem here when `arr` is empty. Please add a special case to return early in that case.
Furthermore, as it currently is, this may (unnecessarily) populate the hash part if `min_key` is not `1`, so please shift accordingly to prevent that.